dali/pipeline/operator/builtin/external_source.h

```diff
@@ -287,6 +297,23 @@ class ExternalSource : public Operator<Backend>, virtual public BatchSizeProvide
   DISABLE_COPY_MOVE_ASSIGN(ExternalSource);
 
  protected:
+  inline bool HasNdim() {
+    return !layout_.empty() || spec_.HasArgument("ndim");
+  }
+
+  inline void InferNdim() {
+    if (!layout_.empty()) {
+      if (ndim_ != -1) {
+        DALI_ENFORCE(ndim_ == layout_.ndim(), make_string("Dimensionality of the provided "
```

Comment:
As above, I feel this would be easier to read. But, again, your call ;)
```suggestion
        DALI_ENFORCE(ndim_ == layout_.ndim(), make_string("Number of dimensions in the provided "
```

Reply:
done